### PR TITLE
Statically link mingw/msvc runtime libraries on Windows

### LINKED ISSUE #976: Test project builds on Windows 11 64-bit but doesn't work in Godot
I tried compiling beta 8 and 9 versions of Godot-cpp and demo project and they compile ok. But extension doesn't work in Godot b8/b9.

I also tried building GDExtensionSummator and got same results.
Here's also link to GDExtensionSummator issue I posted: [Link](https://github.com/paddy-exe/GDExtensionSummator/issues/29)

Everything work OK on linux x64.

The error:
```
  Can't open dynamic library: C:/Users/Bostjan/work/git/godot-cpp/test/demo/bin/libgdexample.windows.template_debug.x86_64.dll, error: Error 126: The specified module could not be found.
  .
  core/extension/gdextension.cpp:400 - GDExtension dynamic library not found: C:/Users/Bostjan/work/git/godot-cpp/test/demo/bin/libgdexample.windows.template_debug.x86_64.dll
  Failed loading resource: res://example.gdextension. Make sure resources have been imported by opening the project in the editor at least once.
  Transient parent has another exclusive child.
  Cannot get class 'Example'.
  Cannot get class 'ExampleMin'.
  In Object of type 'MissingNode': Attempt to connect nonexistent signal 'custom_signal' to callable 'Node(main.gd)::_on_Example_custom_signal'.
```

### LINKED ISSUE #1064: `Can't open dynamic library: [...] error: Error 126: The specified module could not be found.` But the .dll is right there
System: Windows 11
Godot Version:  v4.0.stable.official.92bee43ad
godot-cpp version: [4.0 branch](https://github.com/godotengine/godot-cpp/tree/4.0)
extension source code: [qrnative](https://github.com/fenix-hub/qrnative-cpp)

Last time I tried to compile and use my custom extension worked (it was on Godot 4 beta 3).
Even after updating everything, and following the documentation, when I try to use the .dll in Godot after compilation of both `godot-cpp` and my library, the following message appears in the Editor:
```console
Godot Engine v4.0.stable.official (c) 2007-present Juan Linietsky, Ariel Manzur & Godot Contributors.
  modules/gltf/register_types.cpp:70 - Blend file import is enabled in the project settings, but no Blender path is configured in the editor settings. Blend files will not be imported.
--- Debug adapter server started ---
--- GDScript language server started ---
  Can't open dynamic library: D:/Development/GodotEngine/4.x/qrnative-cpp/project/addons/qrnative/bin/windows/libqrnative.windows.template_debug.x86_64.dll, error: Error 126: The specified module could not be found.
  .
  core/extension/gdextension.cpp:400 - GDExtension dynamic library not found: D:/Development/GodotEngine/4.x/qrnative-cpp/project/addons/qrnative/bin/windows/libqrnative.windows.template_debug.x86_64.dll
  Failed loading resource: res://addons/qrnative/qrnative.gdextension. Make sure resources have been imported by opening the project in the editor at least once.
```
Altough the path to the .dll is correct and there are no typos.
image
Here's the `.gdextension` file
```
[configuration]

entry_symbol = "qrnative_init"

[libraries]

windows.debug.x86_64 = "bin/windows/libqrnative.windows.template_debug.x86_64.dll"
# also tried with "res://addons/qrnative/bin..."
```
I've put the link to the source code, if anybody using w11 want's to try to test the extension. The source code contains the example project used in this issue.

Anyway, [compilation tests](https://github.com/fenix-hub/qrnative-cpp/actions/runs/4331843708/jobs/7563941263) work on all of the desktop platforms, so I don't know if I'm really missing something or it is something on Windows 11.

### PR DESCRIPTION
**Pull request by dsnopek**
This builds on PR https://github.com/godotengine/godot-cpp/pull/1082, implementing the requested review, and removing some changes that seem unrelated to the goal of the PR.

This is still a draft because I haven't yet had a chance to try the MSVC-related changes. However, this is working in my testing with mingw.

Thanks to @feiyunw for making the original PR!

Fixes #1064 
Fixes #976

### CODE REVIEW

File: tools/targets.py [hunk 1]
```diff
@@ -69,6 +69,8 @@ def generate(env):
 
         if env["optimize"] == "debug" or env["optimize"] == "none":
```
Comment (bruvzg) [inline tools/targets.py:70]: Normal debug build probably should use MT / MD based on static flag. The engine has a separate `debug_crt` flag for the MDd which is not used by default (not sure if there's any use for it besides debugging CRT related issues).

File: tools/windows.py [hunk 2]
```diff
@@ -45,6 +46,18 @@ def generate(env):
         env["SHLIBPREFIX"] = ""
         # Want dll suffix
         env["SHLIBSUFFIX"] = ".dll"
+
+        env.Append(CCFLAGS=["-Wwrite-strings"])
+        if env["use_static_cpp"]:
+            env.Append(
+                LINKFLAGS=[
+                    "-static",
+                    "-Wl,--no-undefined",
```
Comment (bruvzg) [inline tools/windows.py:55]: "no undefined" probably should be there for both static and dynamic.

File: tools/targets.py (hunk 1 above)
Comment (akien-mga) [reply tools/targets.py:70]: Yeah we shouldn't use `/MDd` unless there's a strong reason to.

The flags should be kept in sync with upstream Godot ideally:

https://github.com/godotengine/godot/blob/master/SConstruct#L574-L607

https://github.com/godotengine/godot/blob/master/platform/windows/detect.py#L349-L356

File: tools/windows.py (hunk 2 above)
Comment (dsnopek) [reply tools/windows.py]: Makes sense! I moved it up to apply in both cases

File: tools/targets.py (hunk 1 above)
Comment (dsnopek) [reply tools/targets.py]: Ah, ok, thanks! The scope of this PR may be wandering just the tiniest bit :-) but I've gone ahead and synchronized the optimization flags along with the "use_static_cpp" bits from upstream Godot

**Comment by dsnopek** [comment]
Alright, I've tested this with MSVC and it seems to be working! So, I'm taking it out of "draft" now :-)

File: tools/targets.py [hunk 3]
```diff
@@ -60,17 +60,17 @@ def generate(env):
...
-
-        if env["optimize"] == "debug" or env["optimize"] == "none":
-            env.Append(CCFLAGS=["/MDd", "/Od"])
-        else:
-            env.Append(CCFLAGS=["/MD"])
+        elif env["optimize"] == "debug" or env["optimize"] == "none":
+            env.Append(CCFLAGS=["/Od"])
 
     else:
         if env["debug_symbols"]:
```
Comment (akien-mga) [inline tools/targets.py:76]: Another potential sync with Godot, to add here:
```
            # Adding dwarf-4 explicitly makes stacktraces work with clang builds,
            # otherwise addr2line doesn't understand them
            env.Append(CCFLAGS=["-gdwarf-4"])
```
But that can also go in another PR.

**Review by akien-mga — APPROVED** [review]
Looks great!
